Issue 872 UI change for surface geometry searching

juzen2003 opened the pull request:
- Fixes #872
- Fixes #663 
- Run JSHint on: search.js, widgets.js, menu.js, hash.js
- Test on latest: Chrome, Firefox, Safari, and Opera.
- One issue: getting mults hints when switching to a different surfacegeo target:
  -  We not only change surfacegeo target name, but also change all related surfacegeo slugs to new target ones. So the mults hints displayed in surface target name is the correct return from api call but it will not be the result count for the future next click. For example, switching in surfacegeo target name from surfacegeometrytargetname=Mimas&SURFACEGEOmimasplanetographiclatitude1=7 (Pallene has 1 result) to surfacegeometrytargetname=Pallene&SURFACEGEOpalleneplanetographiclatitude1=7 (Pallene actually has 0 result, not 1).

juzen2003 committed:
Add each surfacegeo target slug into the data-slug attribute of the corresponding
surfacegeo radio input.
- In widgets.js (line 755-764), create getSurfacegeoTargetSlug to convert
  surfacegeo pretty name to surfacegeo slug.
- In widgets.js (line 741-748), put each surfacegeo target slug into the
  data-slug attribute of the corresponding radio input.
- In import_util.py (line 265-266), add a comment as a reminder to change
  getSurfacegeoTargetSlug in widgets.js whenever slug_name_for_sfc_target
  function is changed.

juzen2003 committed:
Create oldSurfacegeoTarget to store the current surfacegeo target slug.
This will be used to detect the old surfacegeo related slugs for swapping
with the new surfacegeo related slugs.
- In hash.js (line 644-650), init opus.oldSurfacegeoTarget in initFromHash
  if the reloading URL has surfacegeometrytargetname.
- In opus.js (line 146), declare oldSurfacegeoTarget to detect the old
  surfacegeo related slugs.
- In utils.js (line 114-123), move getSurfacegeoTargetSlug to utils.js.
- In widgets.js and search.js, add debugging console.log, will remove them
  later.
- In import_util.py (line 266), update comments.

juzen2003 committed:
Update surfacegeo attributes in DOMs when surfacegeo target name is changed.
Also properly update selections, extras, and widgets column to have the
correct URL updated.
- In widget.html (line 14), add class="op-widget-title" to widget title.
  This is to update the widget title in place more easily.
- In widgets.js (line 98-198), modify the singlechoice change event handler
  to properly update related surfacegeo attributes in DOMs, and also update
  opus.selections, opus.extras, and opus.prefs.widgets to have the correct
  URL updated.
- In widgets.js (line 228-240), create updateSURFACEGEOattrInPlace to update
  surfacegeo attributes in place with newly select target.

juzen2003 committed:
Update get_triggered_tables to push in surfacegeo target table even if the
it has 0 result. This will make sure newly selected surfacegeo target has
related menu item displayed in search sidebar.
- In views.py (line 1580-1582), push in table to triggered_tables even if
  the newly selected surfacegoe target has 0 result.
- In menu.js, clean up the indentation.

juzen2003 committed:
Clean up commented out codes and console.log in search.js and widgets.js.

juzen2003 committed:
Update the unit record for surfacegeo target related widgets so that after
switching to a new target and changing unit, the values can be properly
converted.
- In widgets.js (line 157-158), update record in opus.currentUnitBySlug
  when a new surfacegeo target is selected.
- In widgets.js (line 204-217), create updateSURFACEGEODataObj to update
  surfacegeo info in data object (opus.selections, opus.extras, and
  opus.currentUnitBySlug) with newly select target.

rfrenchseti commented on opus/application/static_media/js/widgets.js:
```diff
@@ -93,18 +93,71 @@ var o_widgets = {
+        // Update surfacegeo widgets in place if user selects another surfacegeo target.
+        // 1. Update surfacegeo attributes/text in DOMs of all related surfacegeo widgets.
+        // 2. Update selections (opus.selection, opus.extras) & widgets.
         $('#search').on('change', 'input.singlechoice', function() {
```
Don't we need some kind of test here that it's the surface geo selection that's changing? What if we had another radio button widget in the future?

rfrenchseti commented on opus/import/import_util.py:
```diff
@@ -262,6 +262,8 @@ def table_name_for_sfc_target(target_name):
     return encode_target_name(target_name)
 
+# NOTE: whenever we change this function, we will have to change
+# getSurfacegeoTargetSlug in JS codes as well.
```
"code". Also mention what file it's in.

rfrenchseti commented on opus/application/apps/results/views.py:
```diff
@@ -1577,9 +1577,9 @@ def get_triggered_tables(selections, extras, api_code=None):
-                    [r['target_name'].upper() for r in results]):
-                    triggered_tables.append(partable)
+                # If the selected surfacegeo target has no result, we
+                # still want to have related menu item displayed.
```
have THE related menu item

rfrenchseti commented on opus/application/static_media/js/hash.js:
```diff
@@ -641,6 +641,13 @@ var o_hash = {
                     let slugCounter = o_utils.getSlugOrDataTrailingCounterStr(slug);
                     slug = slugNoCounter;
 
+                    // For surfacegeometrytargetname, we have to assign the init value
```
I found this comment somewhat confusing, because I didn't remember that this function was only called on page reload.

How about "If there is a value specified for surfacegeometrytargetname in the URL, we have to initialize opus.oldSurfacegeoTarget with it so that we can properly handle changes to surfacegeometrytargetname in the future."

rfrenchseti commented on opus/application/static_media/js/utils.js:
```diff
@@ -109,7 +109,18 @@ var o_utils = {
-    }
+    },
+
+    getSurfacegeoTargetSlug: function(target) {
```
Maybe put in a comment here referencing the file in the import pipeline. Something like "This function is patterned after slug_name_for_sfc_target in import_util.py".

rfrenchseti commented on opus/application/static_media/js/widgets.js:
```diff
@@ -93,18 +93,71 @@ var o_widgets = {
-        // close opened surfacegeo widget if user select another surfacegeo target
+        // Update surfacegeo widgets in place if user selects another surfacegeo target.
+        // 1. Update surfacegeo attributes/text in DOMs of all related surfacegeo widgets.
+        // 2. Update selections (opus.selection, opus.extras) & widgets.
```
Should be opus.selections

rfrenchseti commented on opus/application/static_media/js/widgets.js:
```diff
@@ -93,18 +93,71 @@ var o_widgets = {
+            o_widgets.updateSURFACEGEODataObj(opus.selections, oldTargetStr, newTargetStr);
+            o_widgets.updateSURFACEGEODataObj(opus.extras, oldTargetStr, newTargetStr);
+
+            // Update widgets column in opus.prefs
```
Is widgets really a "column"? Maybe "Update widgets array"?

rfrenchseti commented on opus/application/static_media/js/widgets.js:
```diff
@@ -135,6 +188,34 @@ var o_widgets = {
+    updateSURFACEGEODataObj: function(dataObj, oldTargetStr, newTargetStr) {
+        /**
+         * Update data object (opus.selections, opus.extras, and opus.currentUnitBySlug)
+         * in place with newly select target.
```
selected

rfrenchseti replied:
Not yet fixed.

rfrenchseti commented on opus/application/static_media/js/widgets.js:
```diff
@@ -738,6 +821,14 @@ var o_widgets = {
                        $(mult_id).next().slideDown("fast");
                    }
                });
+
```
Just for future reference, these few lines above are also where we should fix issue #663 that's assigned to you.

rfrenchseti reviewed (changes requested):
- Didn't we say that if the "surface geometry target name" widget got closed that it would cause all of the other surface geometry widgets to close as well? Popping up a dialog first if any of them are populated with actual search vales?
- I don't actually understand the "issue" you mention about hint results. If I do your example of selecting Mimas, setting the latitude, and then switching to Pallene, I get a Pallene count of 0, which is correct.

juzen2003 committed:
Close all related SURFACEGEO widgets when the surfacegeometrytargetname
widget is closed.
- In views.py (line 1581), update comments.
- In hash.js (line 643-645), update comments.
- In utils.js (line 114), add comments.
- In import_util.py (line 266), update comments.
- In widgets.js (line 220, 161), update comments.
- In widgets.js (line 111), add a if block to check if the singlechoice
  is a surfacegeo target input, if so, we update surfacegeo widgets in place.
- In widgets.js (line 232-247), create closeAllSURFACEGEOWidgets to close
  all SURFACEGEO related widgets.
- In widgets.js (line 101-103), if the closing widget is the surfacegeometrytargetname
  widget, we will also close all the related SURFACEGEO widgets.
- In widgets.js (line 44-48), move isRemovingEmptyWidget from closeWidget
  to o_widgets. This flag is used to determined if we are closing any
  widget with empty input. In a series of closing widget actions, if there
  is one input in a widget with a value, we will make sure a search is
  performed after updating URL. If none of inputs has value, we don't need
  to perform a search.
- In widget.js/opus.js (3 places), remove delete opus.normalizeInputForAllFieldsInProgress[opus.allSlug]
  if normalizedData.reqno < o_search.lastSlugNormalizeRequestNo. This
  solve the page reloading issue if we keep closing lots of widget at one
  time.

juzen2003 committed:
Expand intended target and surfacegeo input group on permalink.
- In views.py (line 293), add mult group label as the data-group attributes
  of ul element (.mult_group).
- In widgets.js (line 838-852), create expandInputGroup to loop through
  targetInputs and expand the group with selected inputs.
- In widgets.js (line 814-825), modify customWidgetBehaviors to call
  expandInputGroup to expand the selected inputs group on permalink.

rfrenchseti commented on opus/application/static_media/js/widgets.js:
```diff
@@ -135,6 +201,51 @@ var o_widgets = {
+            try {
+                $(id).remove();
+            } catch (e) {
+                console.log("error on close widget, id="+id);
```
We have a opus.logError function if you want to leave this in here

rfrenchseti replied:
Not fixed

rfrenchseti commented on opus/application/static_media/js/widgets.js:
```diff
@@ -41,6 +41,12 @@ var o_widgets = {
     // want to perform a search when a widget has just opened.
     isGetWidgetDone: false,
 
+    // This flag is used to determined if we are closing any widget with empty input.
```
to determine.

rfrenchseti commented on opus/application/static_media/js/widgets.js:
```diff
@@ -41,6 +41,12 @@ var o_widgets = {
+    // This flag is used to determined if we are closing any widget with empty input.
+    // In a series of closing widget actions, if there is one input in a widget with a
+    // value, we will make sure a search is performed after updating URL. If none of
+    // inputs has value, we don't need to perform a search.
```
none of the inputs have values

rfrenchseti commented on opus/application/static_media/js/widgets.js:
```diff
@@ -135,6 +201,51 @@ var o_widgets = {
 
+    updateSURFACEGEOattrInPlace: function(targetElement, newSurfacegeoTargetSlug, attribute) {
+        /**
+         * Update surfacegeo attributes in place with newly select target
```
selected

rfrenchseti reviewed (changes requested):
- Existing functionality has broken. Before, if you selected something in the "Planet" field, the appropriate Intended Target and Surface Geo Target categories would expand. This no longer happens. This should also happen when a fresh URL is loaded with a Planet selected.
- What happened to part two of the issue, where there would be new selected metadata columns as the surface geo target changes?
- When the user clicks on the close icon for surface geo target, we had talked about warning the user that this would also close all of the associated widgets and ask the user for confirmation. If the user said no, then don't close the surface geo target widget at all. This confirmation should be asked if any there are any associated search widgets, even if they don't have values in them.

juzen2003 committed:
Merge remote-tracking branch 'upstream/master' into 872_ui_change_for_surface_geometry_searching

juzen2003 committed:
Restore the original behavior that when a planet is selected, the corresponding
input groups in intended target and surfacegeo target are expanded.
- In widgets.js (line 882-891), create expandInputGroupBySelectedPlanet
  to organize the old code to expand the corresponding input groups based
  on the selected planet.
- In widgets.js (line 816-820, 824-831, 840-847), restore old code with
  expandInputGroupBySelectedPlanet.

juzen2003 committed:
Pop out a confirmation modal when closing surfacegeometrytargetname widget
if there are other SURFACEGEO related widgets open. If there are no other
SURFACEGEO related widgets open, there will be no confirmation modal.
- In base.html (line 488-498), add a confirmation modal #op-close-surfacegeo-widgets.
  It will be popped out when closing surfacegeometrytargetname widget
  if there are other SURFACEGEO related widgets open.
- In opus.js (line 837-839), call o_widgets.closeAllSURFACEGEOWidgets()
  to close all surfacegeo related widgets when submit button is hit.
- In widgets.js (line 249-258), re-organize the code and create
  closeAndRemoveWidgetFromDOM to close widget and remove widget elements
  from DOM.
- In widgets.js (line 165-173), update metadata selections (opus.prefs.widgets)
  when surfacegeo target is changed.

juzen2003 committed:
Update metadata selections when surfacegeotarget is changed in search tab
and there are surfacegeo metadata items (related to old target) selected.
- In browse.js (line 1616-1618), in getDataURL, get the hash string from
  getFullHashStr instead of getHash because we want the updated version
  of cols= even if the main URL hasn't been updated yet.
- In browse.js (line 1896-1898), in loadData, increment opus.lastLoadDataRequestNo[view].
  Previously we didn't increment this reqno for dataimages.json api call.
- In widgets.js (line 170-173), update metadata selections when surfacegeotarget
  is changed and there are surfacegeo metadata items (related to old target)
  selected.

rfrenchseti commented on opus/application/apps/ui/templates/ui/base.html:
```diff
@@ -485,6 +485,17 @@ <h3 class='op-header-text'></h3>
         %}
+        {% include "ui/confirm_modal.html" with
+            id_name="op-close-surfacegeo-widgets"
+            confirm_title="Confirm Close All Surfacegeometry Widgets"
```
"Surfacegeometry" and "surfacegeometry" are not very friendly terms for the end user. "Widget" is not a term the user understands. We call them "search fields" in the documentation. Also it won't be clear to them why the widgets are being closed.

Title: "Confirm Close All Surface Geometry Search Fields"

Msg: "Closing the Surface Geometry Target Name search field will also close all target-specific search fields. Are you sure you wish to proceed?"

rfrenchseti reviewed (changes requested):
- When adding the metadata fields, the new fields should appear at the end of the list in the same order as they were originally. So if we have Latitute [Pallene], Longitude [Pallene] and the user selects Epimetheus, we should end up with Latitute [Pallene], Longitude [Pallene], Latitute [Epimetheus], Longitude [Epimetheus]. Feel free to ask about this kind of thing before coding.
- We should also update the getting started guide. Be sure to pull the changes I made to the master branch a few minutes ago first. Here is text to add at line 120:

```
<p>The <b>Surface Geometry Target Name</b> search field is special. Only one
    surface geometry target can be selected at a time. When you select a surface
    geometry target, a new <b>[Target] Surface Geometry Constraints</b>
    category will become available providing search fields for that single
    target. If you later change the surface geometry target, all of
    the existing target-specific fields you selected will automatically switch
    to the new target but keep their existing search values. This is useful
    for cases such as searching for similar lighting geometry for different
    targets. In addition, any metadata fields you selected for the previous
    target will be duplicated for the new target. If you close the
    <b>Surface Geometry Target Name</b>, all target-specific search fields will
    also be closed.</p>
```

juzen2003 committed:
Merge remote-tracking branch 'upstream/master' into 872_ui_change_for_surface_geometry_searching

juzen2003 committed:
Update gettingStarted guide for surfacegeometry field, update the title
and messages in #op-close-surfacegeo-widgets modal, and add the new
surfacegeo metadata items at the end of opus.prefs.col.
- In gettingstarted.html (line 121-133), add new info for surfacegeometry
  field.
- In base.html (line 490-492), update the title and messages of
  #op-close-surfacegeo-widgets modal.
- In widgets.js (line 176-177), append the new surfacegeometry metadata
  selections to the end of opus.prefs.col.

juzen2003 committed:
Update the error message when and error occurs at widget closing event.
- In widdgets.js (line 262), update the error message.

juzen2003 committed:
When there is any value in SURFACEGEO related widgets, in surfacegeometrytargetname
, we will have "--" next to non-selected inputs and green hint number next
to the selected input. When there is no value in SURFACEGEO related widgets,
in surfacegeometrytargetname, we will have green hints (numbers) next to all inputs.
- In opus.js (line 375-379), update the logic to have "--" next to non-selected
  inputs and green hint number next to the selected input in surfacegeometrytargetname
  when there are values in SURFACEGEO related widgets.
- In search.js (line 1155-1181), add a parameter hideHintsForSurfacegeoTarget
  to getValidMults, if it's true, have "--" next to non-selected inputs
  and green hint number next to the selected input in surfacegeometrytargetname.

juzen2003 committed:
Fix the comment in opus.js (line 373).

rfrenchseti reviewed (changes requested):
- If you click on radio buttons rapidly, you can get the selected metadata list to screw up. This is especially true with "Pan" and "Pandora".
- The "Surface Geometry Target Name" field should never show up in the "Current Search Terms" part of select metadata.
- It's possible to-deselect all radio buttons by clicking just to the left of the selected button.
